ES|QL: Check if cluster aliases and index patterns are valid before executing query

pawankartik-elastic opened the pull request:
Consider the queries with more than 1 index seperator (not valid as per our current spec):

1. `FROM remote:"foo:bar"`, and,
2. `FROM "remote:foo:bar:baz"`

As of now, both the queries return empty results when `POST`ing to `/_query?format=txt`. This happens because they're syntactically valid (but not semantically) and the subsequent validation makes no attempt to discard them or error out. This PR fixes this behaviour by failing such queries at the parsing stage. There was an initial concern surrounding the licensing and security aspects but since we're failing fast at the parsing stage and aren't allowing the queries "through", we should be fine. Once malformed queries are discarded, subsequent queries should go through the validation steps fine.

Addresses https://github.com/elastic/elasticsearch/issues/121901.

Considerations:

1. ~~Should the `::` operator be accounted for (because the respective PR is yet to be merged)?~~ Changes have been pulled into my branch and accommodated.
2. ~~This is a bug fix. How far should it be backported (keeping in mind the `::` operator)?~~ Looking good to go back as far as 8.19.
3. ~~What's the right team to tag? SF/CCS or Analytics/ES|QL?~~

pawankartik-elastic committed:
Check if index patterns conform to valid format before validation

pawankartik-elastic committed:
Update docs/changelog/122497.yaml

elasticsearchmachine commented:
Hi @pawankartik-elastic, I've created a changelog YAML for you.

pawankartik-elastic committed:
Let `validateClusterString()` look for `REMOTE_CLUSTER_INDEX_SEPARATOR`
in cluster name

pawankartik-elastic committed:
Merge branch 'main' into pkar/index-pattern-check

pawankartik-elastic commented:
Things unclear to me at the moment:

1. Why are we skipping validation when remote is specified?
2. Is there a single API we could use everywhere to parse and validate remote and index names?
3. Why do we even allow quoting in the patterns (specified in the grammar file)?

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -78,6 +78,15 @@ public String visitIndexPattern(List<EsqlBaseParser.IndexPatternContext> ctx) {
+            }
             // skip validating index on remote cluster, because the behavior of remote cluster is not consistent with local cluster
             // For example, invalid#index is an invalid index name, however FROM *:invalid#index does not return an error
             if (clusterString == null) {
```
This shouldn't be an if-else, the index pattern should always be validated.

astefan replied:
CC @fang-xing-esql. Maybe you have more context behind this decision.

fang-xing-esql replied:
Currently, we don't validate index names in `IdentifierBuilder` for remote clusters, when I implemented the validation of index names, I found the behaviors of invalid index names between remote and local cluster are not quite consistent, so I decide to skip the validation of index names on remote cluster.

For example, these queries below all run to completion, they don't error out although the remote index names are invalid

```
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{ 
  "query": "FROM existing_index, *:\"invalid:index\""
}
'
        `*        |    `*.keyword    |     author      | author.keyword  |     b`a`r     |      bar      |  page_count   |      release_date      
------------------+------------------+-----------------+-----------------+---------------+---------------+---------------+------------------------
Snow Crash        |Snow Crash        |Neal Stephenson  |Neal Stephenson  |5              |1              |470            |1992-06-01T00:00:00.000Z
Revelation Space  |Revelation Space  |Alastair Reynolds|Alastair Reynolds|4              |2              |585            |2000-03-15T00:00:00.000Z
1984              |1984              |George Orwell    |George Orwell    |3              |3              |328            |1985-06-01T00:00:00.000Z
Fahrenheit 451    |Fahrenheit 451    |Ray Bradbury     |Ray Bradbury     |2              |4              |227            |1953-10-15T00:00:00.000Z
Brave New World   |Brave New World   |Aldous Huxley    |Aldous Huxley    |1              |5              |268            |1932-06-01T00:00:00.000Z
The Handmaids Tale|The Handmaids Tale|Margaret Atwood  |Margaret Atwood  |0              |0              |311            |1985-06-01T00:00:00.000Z
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{ 
  "query": "FROM existing_index, \"*:invalid:index\""
}
'
        `*        |    `*.keyword    |     author      | author.keyword  |     b`a`r     |      bar      |  page_count   |      release_date      
------------------+------------------+-----------------+-----------------+---------------+---------------+---------------+------------------------
Snow Crash        |Snow Crash        |Neal Stephenson  |Neal Stephenson  |5              |1              |470            |1992-06-01T00:00:00.000Z
Revelation Space  |Revelation Space  |Alastair Reynolds|Alastair Reynolds|4              |2              |585            |2000-03-15T00:00:00.000Z
1984              |1984              |George Orwell    |George Orwell    |3              |3              |328            |1985-06-01T00:00:00.000Z
Fahrenheit 451    |Fahrenheit 451    |Ray Bradbury     |Ray Bradbury     |2              |4              |227            |1953-10-15T00:00:00.000Z
Brave New World   |Brave New World   |Aldous Huxley    |Aldous Huxley    |1              |5              |268            |1932-06-01T00:00:00.000Z
The Handmaids Tale|The Handmaids Tale|Margaret Atwood  |Margaret Atwood  |0              |0              |311            |1985-06-01T00:00:00.000Z
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{ 
  "query": "FROM existing_index, *:\"-invalid:index\""
}
'
        `*        |    `*.keyword    |     author      | author.keyword  |     b`a`r     |      bar      |  page_count   |      release_date      
------------------+------------------+-----------------+-----------------+---------------+---------------+---------------+------------------------
Snow Crash        |Snow Crash        |Neal Stephenson  |Neal Stephenson  |5              |1              |470            |1992-06-01T00:00:00.000Z
Revelation Space  |Revelation Space  |Alastair Reynolds|Alastair Reynolds|4              |2              |585            |2000-03-15T00:00:00.000Z
1984              |1984              |George Orwell    |George Orwell    |3              |3              |328            |1985-06-01T00:00:00.000Z
Fahrenheit 451    |Fahrenheit 451    |Ray Bradbury     |Ray Bradbury     |2              |4              |227            |1953-10-15T00:00:00.000Z
Brave New World   |Brave New World   |Aldous Huxley    |Aldous Huxley    |1              |5              |268            |1932-06-01T00:00:00.000Z
The Handmaids Tale|The Handmaids Tale|Margaret Atwood  |Margaret Atwood  |0              |0              |311            |1985-06-01T00:00:00.000Z
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{     
  "query": "FROM existing_index, \"*:-invalid:index\""
}     
'
        `*        |    `*.keyword    |     author      | author.keyword  |     b`a`r     |      bar      |  page_count   |      release_date      
------------------+------------------+-----------------+-----------------+---------------+---------------+---------------+------------------------
Snow Crash        |Snow Crash        |Neal Stephenson  |Neal Stephenson  |5              |1              |470            |1992-06-01T00:00:00.000Z
Revelation Space  |Revelation Space  |Alastair Reynolds|Alastair Reynolds|4              |2              |585            |2000-03-15T00:00:00.000Z
1984              |1984              |George Orwell    |George Orwell    |3              |3              |328            |1985-06-01T00:00:00.000Z
Fahrenheit 451    |Fahrenheit 451    |Ray Bradbury     |Ray Bradbury     |2              |4              |227            |1953-10-15T00:00:00.000Z
Brave New World   |Brave New World   |Aldous Huxley    |Aldous Huxley    |1              |5              |268            |1932-06-01T00:00:00.000Z
The Handmaids Tale|The Handmaids Tale|Margaret Atwood  |Margaret Atwood  |0              |0              |311            |1985-06-01T00:00:00.000Z
```

If we decide to validate remote index names, it does require more tests so that we are clear about the scope of change. `:` is not the only invalid character, and also be careful about `*` and `-` in the (remote)index patterns `IndentifierBuilder.validateIndexPattern` calls some existing APIs to validate index names, and remote cluster index name validation can take advantage of it. It could change the current behaviors related to (remote) index patterns potentially, we need to justify if the behavior changes make sense.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
I think we need a bunch more tests for invalid index pattern in all the cases where there is a remote prefix `some_remote:...`. Ideally, randomly generated invalid index patterns.

alex-spies reviewed (changes requested):
Heya, thanks @pawankartik-elastic !

1. This approach is unfortunately insufficient, but not by much.

This PR improves the situation for the buggy queries that initially triggered the issue. But it doesn't solve the problem that the index pattern may be otherwise invalid, like in `FROM valid_remote:"invalid<index_name"` - which still just returns 
```
  <no-fields>  
---------------
```

The reason is that in `IdentifierBuilder.visitIndexPattern`, we just never call `validateIndexPattern` in case that `clusterString` is not null. Additionally, inside `validateIndexPattern`, we explicitly skip the validation in case that the pattern contains a `:`.

I believe the correct solution should try to split the `indexPattern` at the first `:` in case that there is no `clusterString` and this is not a date math index pattern. And then, the remaining, actual index pattern should be sent to `validateIndexPattern`, which needs additionally to consider the `:` character as invalid rather than just skipping validation.

2. I think there is a separate problem where cluster names can be invalid, which we don't reject at parse time, but in this case we just throw `no_such_remote_cluster_exception`.

This may be cheap enough to not be a problem - although we could improve `validateIndexPattern` too check for correct remote name syntax, too.

3. On a separate note: here https://github.com/elastic/elasticsearch/pull/120355/files#diff-ef7f022d55c8068ca9c7c26cc5bac2c4cc3cf5689a9bfc20e931cdf17541556eL83-L84 Ievgen added randomly generated cluster names to our parser tests.

I think it'd be nice to go and check if they use all possible characters that can be used in remote names? If there are more characters than that, we should go and make the identifier generator use that, so this is included in our tests. For instance, what about unicode? (I don't know :) )

pawankartik-elastic commented:
Thank you @alex-spies. I agree that this fix doesn't fully address the issue. I was simply trying to:

1. Validate my approach (and kick-off some internal discussions with the CCS team; will document them briefly here before the PR merges for readers' context), and,
2. Discuss a few uncertainities (of why certain things were done certain way).

I'll now work on ensuring that we fully address the issue.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -138,4 +147,20 @@ private static void validateIndexPattern(String indexPattern, EsqlBaseParser.Ind
+            } else if (ch == '>') {
+                inDateTime = false;
+            } else if (ch == REMOTE_CLUSTER_INDEX_SEPARATOR && inDateTime == false) {
+                seperatorsCount++;
```
There is no need to go through the entire loop and wait until the end to return `seperatorsCount > maxAllowedSeparators`. Do this comparison here, after the increment, and return early if `seperatorsCount > maxAllowedSeparators`. Otherwise, after the loop, `return false`.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -78,6 +78,15 @@ public String visitIndexPattern(List<EsqlBaseParser.IndexPatternContext> ctx) {
+            // other aspects of them are valid. There's no point in running the second check if the first one fails.
+            // Should clusterString be non-null, validateClusterString() handles the required validation. So for now, look only at
+            // the indexPattern.
+            var maxSeparators = clusterString == null ? 1 : 0;
+            if (patternExceedsMaxIndexSeparators(indexPattern, maxSeparators)) {
+                throw new ParsingException(source(c), "Unexpected index separator in index pattern");
+            }
```
All this code can be executed before `visitClusterString` above.
Also, since the index separators check is part of the indexPattern validation, why is this code not part of `validateIndexPattern` method?

pawankartik-elastic replied:
1. In this revision, `validateIndexPattern()` is only called when `clusterString` is `null`. I want this specific code to execute irrespective of that.
2. This specific code relies on whether `clusterString`'s `null`-ness to decide the max allowed separators. If I moved it to before `visitClusterString()`, I wouldn't know if it'd be `null` or not, right?

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -138,4 +147,20 @@ private static void validateIndexPattern(String indexPattern, EsqlBaseParser.Ind
+        int seperatorsCount = 0;
+        boolean inDateTime = false;
+        for (char ch : pattern.toCharArray()) {
+            if (ch == '<') {
```
Regarding this check for `<` and `>` (since I don't know all the fine details of ES validation), can you point us to what you mentioned in the description of the PR - our spec around these restrictions?

> more than 1 index seperator (not valid as per our current spec)

astefan reviewed (changes requested):
Apart from what @alex-spies awesomely mentioned already, please add many many more tests. For one, there is no test with date math expressions (something you explicitly consider in your code).

pawankartik-elastic committed:
Merge branch 'main' into pkar/index-pattern-check

pawankartik-elastic committed:
Mute 3 cases till further clarification and fix error string message

pawankartik-elastic committed:
Fix tests

pawankartik-elastic committed:
Fix bug in breaking down indices

pawankartik-elastic committed:
Merge branch 'main' into pkar/index-pattern-check

pawankartik-elastic committed:
Tiny refactoring around how wildcard is processed and added tests

Commit:
[CI] Auto commit changes from spotless

pawankartik-elastic committed:
Drop duplicated test cases and fix flaky-ness caused by quoting

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

pawankartik-elastic commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -784,6 +781,8 @@ public void testInvalidQuotingAsFromIndexPattern() {
 
         expectError("FROM \"\"\"foo\"\"\"bar\"\"\"", ": mismatched input 'bar' expecting {<EOF>, '|', ',', 'metadata'}");
         expectError("FROM \"\"\"foo\"\"\"\"\"\"bar\"\"\"", ": mismatched input '\"bar\"' expecting {<EOF>, '|', ',', 'metadata'}");
+        expectError("FROM remote:\"foo:bar\"", "Index pattern [foo:bar] contains a cluster alias despite specifying one [remote]");
```
Ignore these 2 scenarios. They're already tested below; will drop them soon.

pawankartik-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -148,35 +128,103 @@ protected static void validateClusterString(String clusterString, EsqlBaseParser
+                if (selectorString != null) {
+                    throwOnMixingSelectorWithCluster(reassembleIndexName(clusterString, indexPattern, selectorString), ctx);
+                }
+                validateClusterString(clusterString, ctx);
```
Do we allow unicode chars in the cluster name? Are the naming rules documented anywhere?

pawankartik-elastic committed:
Set cluster string to `null` when it cannot be associated with an index
pattern

pawankartik-elastic committed:
Generate correct invalid patterns

pawankartik-elastic committed:
Merge branch 'main' into pkar/index-pattern-check

fang-xing-esql commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -694,20 +692,19 @@ public void testInvalidCharacterInIndexPattern() {
-            clustersAndIndices(command, "index*", "-index#pattern");
-            clustersAndIndices(command, "*", "-<--logstash-{now/M{yyyy.MM}}>");
-            clustersAndIndices(command, "index*", "-<--logstash#-{now/M{yyyy.MM}}>");
+            expectInvalidIndexNameErrorWithLineNumber(command, "*, index#pattern", lineNumber, "index#pattern", "must not contain '#'");
```
This changes the current behavior. Today, if a `*` exists in index patterns, invalid index names are ignored, they don't error out as far as the index patterns match at least one index(valid and existing). The same applies to the tests a few lines below that have selectors. However, it looks reasonable to me that as soon as there is an index pattern with invalid character found, we error out. In this situation we have a `*`, that means there is a high chance that this whole index pattern in the `from` command matches an existing index, and it doesn't look completely wrong to tolerate invalid character(could be a typo) either. There could be a reason that why this was allowed originally.

pawankartik-elastic replied:
Understood. Should we have a slightly detailed discussion internally with the ES|QL and the CCS team regarding whether we should tolerate such erroneous patterns?

fang-xing-esql replied:
It changes the current behaviors, it worths bringing up to a broader group to discuss further in my opinion, and also add `ES|QL-ui` label to inform kibana team.

pawankartik-elastic replied:
This behaviour is about excluding certain patterns as long as there's a preceding pattern(s) with a `*` (and does not require any valid and existing indices). Since this functionality existed as a way to exclude arbitrary patterns (but is unfortunately not documented anywhere?), I've reverted some of my changes to retain this behaviour and added back these test cases.

fang-xing-esql reviewed (commented):
Thank you @pawankartik-elastic! I added some thoughts.

pawankartik-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -87,33 +88,12 @@ public String visitSelectorString(EsqlBaseParser.SelectorStringContext ctx) {
-                }
-            }
+
+            validateClusterAndIndexPatterns(indexPattern, c, clusterString, selectorString);
```
`validateClusterAndIndexPatterns()` is a `static` method currently. If I were to pass the `IndexPatternContext`, I'd have to make it non-`static` since the visitor methods (such as `visitIndexString()`) are non-`static` and are overridden. Would that be alright with you?

pawankartik-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -148,35 +128,109 @@ protected static void validateClusterString(String clusterString, EsqlBaseParser
+        String[] patterns = indexPattern.split(",");
+        boolean isFirstPattern = true;
+
+        for (String pattern : patterns) {
```
Isn't that exactly what we're doing in this loop? At this point, we loop over each pattern and then after line 195, we validate the `clusterString` (if set), then call the `validateIndicesForCluster()` to validate the indices specified, and then we finally validate the `selectorString` via `validateIndexSelectorString()`.

I maintained the same order. Had I deviated from this order, wrong error messages would have taken precedence and caused the tests to fail due to incorrect error strings (which indeed happened when I was working on this PR—when I deviated from the previous validation ordering).

pawankartik-elastic replied:
Resolving this comment as we've made the grammar changes and this code is no longer relevant/exists.

pawankartik-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -148,35 +128,109 @@ protected static void validateClusterString(String clusterString, EsqlBaseParser
+                indices[0] = clusterAliasAndIndex[1];
+
+                /*
+                 * What if the pattern is index1|index2? We cannot split at the pipe char blindly as that'd mess with
```
We do have a couple of tests for this:
```
// You can use | for remote indices.
assertStringAsIndexPattern("cluster:index|pattern", ...);
assertStringAsIndexPattern("*:index|pattern", ...);

// Error because you cannot use | for local indices.
expectInvalidIndexNameErrorWithLineNumber(command, "index|pattern", lineNumber);
```

I can add a few more. Sounds good?

pawankartik-elastic replied:
I'm resolving this comment. I was mistaken that pipe char was a valid char/sep. Since it's not, I've removed this splitting.

pawankartik-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -148,35 +128,109 @@ protected static void validateClusterString(String clusterString, EsqlBaseParser
                 // throws exception if the selector expression is invalid. Selector resolution does not complain about exclusions
                 throw new ParsingException(e, source(ctx), e.getMessage());
             }
-            hasSeenStar = index.contains(WILDCARD) || hasSeenStar;
```
This is related to your comment here: https://github.com/elastic/elasticsearch/pull/122497/files/7da29d8e2aa80557c1102168e686768647969a0f#r2043198666—at present, if there's a `*`, a valid index, and an invalid pattern, we tolerate it. The `hasSeenStar` used to be passed as an argument to track if we've previously seen `*` and then is used to make a decision—whether to tolerate an invalid pattern or not.

pawankartik-elastic replied:
https://github.com/elastic/elasticsearch/pull/122497#discussion_r2046997133

I've reverted some of my changes to allow exclusion patterns, inline with previous behaviour and that addresses this conversation.

pawankartik-elastic commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -488,15 +489,12 @@ public void testStringAsIndexPattern() {
             clusterAndIndexAsIndexPattern(command, "cluster:.index");
             clusterAndIndexAsIndexPattern(command, "cluster*:index*");
-            clusterAndIndexAsIndexPattern(command, "cluster*:<logstash-{now/D}>*");// this is not a valid pattern, * should be inside <>
-            clusterAndIndexAsIndexPattern(command, "cluster*:<logstash-{now/D}*>");
```
Previously, this would not error out even if the logstash part was invalid because it's a remote index and we use to skip validation for remote indices. Now that we validate even remote indices, these cases should error out with a message along the lines of 'D' is an invalid unit. Such an exact test case is already present so these are duplicates.

pawankartik-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -198,9 +252,44 @@ private static void validateIndexPattern(String indexPattern, EsqlBaseParser.Ind
         }
     }
 
+    private static String[] breakPatternIntoIndices(String pattern) {
```
I don't think `_search` is making any similar validation attemps. I tried `GET`ting `/remote*:foo|bar/_search` and got back a response with this fragment:

```
{
    "error": {
        "root_cause": [
            {
                "type": "index_not_found_exception",
                "reason": "no such index [foo|bar]",
                "index_uuid": "_na_",
                "resource.type": "index_or_alias",
                "resource.id": "foo|bar",
                "index": "foo|bar"
            }
        ],
        "type": "index_not_found_exception",
        "reason": "no such index [foo|bar]",
        "index_uuid": "_na_",
        "resource.type": "index_or_alias",
        "resource.id": "foo|bar",
        "index": "foo|bar"
    },
    "status": 404
}
```

If a similar validation existed, I don't think the query would have even been dispatched.

`resolveDateMathExpression()` looks primarily for the positions of `{` and `}` and then does the replacing, albeit taking the escape chars into account.

pawankartik-elastic replied:
Resolving this conversation: https://github.com/elastic/elasticsearch/pull/122497#discussion_r2046984779

pawankartik-elastic committed:
Address review comments and don't break indices into its constituents

1. Wrapped `splitIndexName()` in try catch and ensured it throws parse
   exception.
2. Relaxed the validation to allow excluding garbage.
3. Validation no longer breaks patterns on pipe char as it's an invalid
   char.

pawankartik-elastic commented:
Ignore the CI failures. It's mostly BWC failures because branch is out of date.

alex-spies reviewed (commented):
Thanks a lot @pawankartik-elastic ! This is looking much better already.

I reviewed the production code, but not yet the new test that was added. The reason is that there have been important discussions on what should and should not be allowed in `FROM` commands, especially the comments by @jbaiera [here](https://github.com/elastic/elasticsearch/issues/122651#issuecomment-2669645128) and [here](https://github.com/elastic/elasticsearch/issues/122651#issuecomment-2672282739), which are not reflected in this PR.

If we go with the suggestions by @jbaiera , and I believe we should, then the present PR could be simplified.

Changing the validation of the `FROM` command is **high risk**, as any potentially introduced new bug that we don't happen to already have tests for can and will break customer queries immediately as this sees the light of day in Serverless.

Therefore, I'd like to double check if we can go with a simpler approach based on @jbaiera 's suggestion. That would IMO likely lead to simpler validation code and thus require less paranoid double checking during review to have good confidence in the change.

(I'm not saying the current code likely has bugs, just that I, as a human, am likely to overlook bugs in code that requires multiple nested loops handling quoted and unquoted index patterns and mutating the current cluster string and so on.)

pawankartik-elastic committed:
Merge branch 'main' into pkar/index-pattern-check

pawankartik-elastic committed:
Adhere to the new grammar

pawankartik-elastic committed:
Merge branch 'main' into pkar/index-pattern-check

pawankartik-elastic committed:
Update docs/changelog/122497.yaml

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -494,18 +494,29 @@ public void testStringAsIndexPattern() {
             clusterAndIndexAsIndexPattern(command, "cluster:index");
             clusterAndIndexAsIndexPattern(command, "cluster:.index");
             clusterAndIndexAsIndexPattern(command, "cluster*:index*");
-            clusterAndIndexAsIndexPattern(command, "cluster*:<logstash-{now/D}>*");// this is not a valid pattern, * should be inside <>
```
Reiterating the previous review: why is this removed? I tested this index pattern and I get this `org.elasticsearch.ElasticsearchParseException: unit [D] not supported for date math [/D]`. This means that the index pattern is valid from ESQL pov, but incorrect from ES point of view. Also, the comment on the same line is incorrect according to my tests. Which means, imo, this should be `from cluster*:<logstash-{now/d}>*` (**notice the small capital `d`**) and this one passes. So, this test should be, at most, **slightly changed**, not removed,

pawankartik-elastic replied:
Done! I've added this case back.

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -494,18 +494,29 @@ public void testStringAsIndexPattern() {
             clusterAndIndexAsIndexPattern(command, "cluster:.index");
             clusterAndIndexAsIndexPattern(command, "cluster*:index*");
-            clusterAndIndexAsIndexPattern(command, "cluster*:<logstash-{now/D}>*");// this is not a valid pattern, * should be inside <>
-            clusterAndIndexAsIndexPattern(command, "cluster*:<logstash-{now/D}*>");
```
My comment from above test is valid here as well.

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -494,18 +494,29 @@ public void testStringAsIndexPattern() {
                 lineNumber,
                 "mismatched input '\"index|pattern\"' expecting UNQUOTED_SOURCE"
             );
```
@pawankartik-elastic this is not about your PR, but something is missing from this test method. ES|QL also has `TS` as a source command. See here https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4#L41.

Even if this has nothing to do with the purpose of this PR, I would like this method to be changed to include at least `TS` source command (of course, with the right capability).

pawankartik-elastic replied:
I think it already does? At the beginning of this method, you have:
```java
if (Build.current().isSnapshot()) {
    commands.add("TS");
}
```

So all the test cases in this method are tested for both `FROM` and `TS`. Or did I understand you wrong?

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -3157,6 +3183,87 @@ public void testValidJoinPattern() {
+            // Select an invalid char to sneak in.
+            // Note: some chars like '|' and '"' are excluded to generate a proper invalid name.
+            char[] invalidChars = { ' ', '/', '<', '>', '?' };
+            var randomInvalidChar = invalidChars[randomIntBetween(0, invalidChars.length - 1)];
```
```suggestion
            var randomInvalidChar = randomFrom(invalidChars);
```

astefan reviewed (commented):
I've left some comments.
This is a very sensitive and tricky area we are changing now and **tests are essential** to our confidence in pushing forward this change. Apart from the random tests, I would like to see many more concrete tests:
```
FROM  "*: ,*,"
FROM ",,, "
FROM ",,,"
FROM ", ,," 
FROM ",,,",*
FROM "",*
FROM ""
FROM "<logstash-{now/d{yyyy.MM.dd|+12:00}}>"
FROM "<a|b>"
FROM "*:<logstash-{now/M{yyyy.MM}}>"
```

pawankartik-elastic committed:
Update x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java

Co-authored-by: Andrei Stefan <astefan@users.noreply.github.com>

pawankartik-elastic commented:
> I've left some comments. This is a very sensitive and tricky area we are changing now and **tests are essential** to our confidence in pushing forward this change. Apart from the random tests, I would like to see many more concrete tests:
> 
> ```
> FROM  "*: ,*,"
> FROM ",,, "
> FROM ",,,"
> FROM ", ,," 
> FROM ",,,",*
> FROM "",*
> FROM ""
> FROM "<logstash-{now/d{yyyy.MM.dd|+12:00}}>"
> FROM "<a|b>"
> FROM "*:<logstash-{now/M{yyyy.MM}}>"
> ```

Some of these tests require an explicit fix. From our conversation this afternoon, you said that the issue is different enough to not fix it in this PR. How would you like me to proceed?

pawankartik-elastic committed:
Apply suggestions from review

pawankartik-elastic committed:
Merge branch 'main' into pkar/index-pattern-check

alex-spies reviewed (approved):
I only reviewed the productive changes as Andrei already made suggestions to the tests.

The production code looks right to me; it's also much simpler than the iteration before the grammar change and looks quite clean to me. Thank you!

I spotted minor issues with trailing commas and white spaces, but nothing dramatic. Please consider this PR unblocked and ready to merge once @astefan approves it.

pawankartik-elastic committed:
Apply suggestions from code review

Co-authored-by: Alexander Spies <alexander.spies@elastic.co>

pawankartik-elastic committed:
Apply suggestions from review

pawankartik-elastic commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/IdentifierBuilder.java:
```diff
@@ -144,59 +129,184 @@ protected static void validateClusterString(String clusterString, EsqlBaseParser
+        if (patterns.length == 0) {
+            throwInvalidIndexNameException(indexPattern, BLANK_INDEX_ERROR_MESSAGE, ctx);
+        } else if (patterns.length == 1) {
+            // Pattern is an unquoted string.
```
Right. The next push fixes the comment to correctly reflect the scenarios. Thanks!

pawankartik-elastic committed:
Merge branch 'main' into pkar/index-pattern-check

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -494,18 +494,29 @@ public void testStringAsIndexPattern() {
-            assertStringAsIndexPattern("*:index|pattern", command + " \"*:index|pattern\"");
+            // Entire index pattern is quoted. So it's not a parse error but a semantic error where the index name
+            // is invalid.
+            expectError(command + " \"*:index|pattern\"", "Invalid index name [index|pattern], must not contain the following characters");
```
@pawankartik-elastic thank you for adding more tests.
What is our plan for this error message?

pawankartik-elastic replied:
This error message comes from `MetadataCreateIndexService.validateIndexOrAliasName()`. One idea is to trap it specifically and re-throw with a different message.

astefan replied:
Yes, please take out the wildcard from there. Thanks.

astefan reviewed (approved):
Apart from the leftover [here](https://github.com/elastic/elasticsearch/pull/122497#discussion_r2166484372), LGTM. Thank you!

pawankartik-elastic committed:
Do not mention asterisk as an invalid char

pawankartik-elastic committed:
Merge branch 'main' into pkar/index-pattern-check
